### PR TITLE
[CodeCamp #18] add ops bbox_overlaps

### PR DESCRIPTION
**Pull request by enemy1205**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add CPU implementation of ops `bbox_overlaps`

cpu version
截图 2023-01-14 10-18-47

pure python version
截图 2023-01-14 10-18-13
## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

No

## Use cases (Optional)

passed `tests/test_ops/test_bbox.py`

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by grimoire** [comment]
Since we already have a pure PyTorch implementation, could you provide a benchmark for the speed-up of the new op?

**Comment by enemy1205** [comment]
> Since we already have a pure PyTorch implementation, could you provide a benchmark for the speed-up of the new op?

I tested it using pytest-benchmark as follows :
Firstly ,in file `tests/test_ops/test_bbox.py` , since the two calls are the same function, only the parameters are different, I commented out the semi-precision call in order to prevent it from averaging

```
    @pytest.mark.parametrize('device', [
        'cpu',
        pytest.param(
            'cuda',
            marks=pytest.mark.skipif(
                not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
        pytest.param(
            'mlu',
            marks=pytest.mark.skipif(
                not IS_MLU_AVAILABLE, reason='requires MLU support')),
        pytest.param(
            'mps',
            marks=pytest.mark.skipif(
                not IS_MPS_AVAILABLE, reason='requires MPS support'))
    ])
    def test_bbox_overlaps_float(self, device,benchmark):
        benchmark(self._test_bbox_overlaps,device, dtype=torch.float)

    # @pytest.mark.parametrize('device', [
    #     pytest.param(
    #         'cuda',
    #         marks=pytest.mark.skipif(
    #             not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
    #     pytest.param(
    #         'mlu',
    #         marks=pytest.mark.skipif(
    #             not IS_MLU_AVAILABLE, reason='requires MLU support'))
    # ])
    # def test_bbox_overlaps_half(self, device,benchmark):
    #     benchmark(self._test_bbox_overlaps,device, dtype=torch.half)
```
Then ,in file `mmcv/ops/bbox.py`
In the end of the function, I implemented calls to  the C++ and pure pytorch versions by commenting out the other one in turn
```
    mode_dict = {'iou': 0, 'iof': 1}
    assert mode in mode_dict.keys()
    mode_flag = mode_dict[mode]
    # Either the boxes are empty or the length of boxes' last dimension is 4
    assert (bboxes1.size(-1) == 4 or bboxes1.size(0) == 0)
    assert (bboxes2.size(-1) == 4 or bboxes2.size(0) == 0)
    assert offset == 1 or offset == 0

    rows = bboxes1.size(0)
    cols = bboxes2.size(0)
    if aligned:
        assert rows == cols
        ious = bboxes1.new_zeros(rows)
    else:
        ious = bboxes1.new_zeros((rows, cols))

    if rows * cols == 0:
        return ious
    
    # Pure Pytorch
    # return _bbox_overlaps_cpu(
    #         bboxes1, bboxes2, mode=mode, aligned=aligned, offset=offset)
    
    # C++ version
    ext_module.bbox_overlaps(
        bboxes1, bboxes2, ious, mode=mode_flag, aligned=aligned, offset=offset)
    return ious
```
By executing `pytest tests/test_ops/test_bbox.py`
The test results are as follows:
image
The top is the C++ version and the bottom is the pure Pytorch version, so there is some speed optimization

File: mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            single_box_iou<T>(boxes1[i].data_ptr<T>(), boxes2[j].data_ptr<T>(),
+                              offset, mode_flag);
+      }
+    }
+  }
+}
+
+void bbox_overlaps_cpu(const Tensor boxes1, const Tensor boxes2, Tensor ious,
+                       const int mode, const bool aligned, const int offset) {
+  bbox_overlaps_cpu_kernel<float>(boxes1, boxes2, ious, mode, aligned, offset);
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp:67]: Assert the input and output datatype here.

File: tests/test_ops/test_bbox.py [hunk 2]
```diff
@@ -34,6 +34,14 @@ def _test_bbox_overlaps(self, device='cpu', dtype=torch.float):
         out = bbox_overlaps(b1, b2, offset=1)
         assert np.allclose(out.cpu().numpy(), should_output, 1e-2)
 
+        b1 = torch.tensor([[1.0 + i, 1.0 + i, 3.0 + i, 3.0 + i]
```
[the remark this answers is not in the document]
Comment (enemy1205) [reply tests/test_ops/test_bbox.py:37]: Sorry, maybe I skipped cuda test before since I only setup cpu environment,do you have any suggestions or I'll just delete this code?

File: mmcv/ops/bbox.py [hunk 3]
```diff
@@ -106,6 +106,28 @@ def bbox_overlaps(bboxes1: torch.Tensor,
 
     rows = bboxes1.size(0)
     cols = bboxes2.size(0)
+    # if aligned:
```
Comment (grimoire) [inline mmcv/ops/bbox.py:109]: Please remove all commented code.

File: mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
+// Copyright(c) OpenMMLab.All rights reserved.
+#include "pytorch_cpp_helper.hpp"
+#include "pytorch_device_registry.hpp"
+#define HOST_DEVICE
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp:4]: What's this for?

**Review by grimoire — APPROVED** [review]
LGTM

File: mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp (hunk 4 above)
Comment (enemy1205) [reply mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp]: I'm sorry, maybe I copied and pasted it up by mistake

Comment (enemy1205) [reply mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp]: I improved my test case. cuda didn't pass the semi-precision before, which should be because my range was too large and I added the previous smaller integer, resulting in the loss of precision. Therefore, I reduced the range and increased the addition number
```
        b1 = torch.tensor([[10.0+i, 10.0+i, 30.0+i, 30.0+i] for i in range(1000)]).to(device).type(dtype)
        b2 = torch.tensor([[20.0+i, 20.0+i, 40.0+i, 40.0+i] for i in range(1000)]).to(device).type(dtype)
        should_output = np.array([1/7]*1000)
        out = bbox_overlaps(b1, b2, aligned=True)
        assert np.allclose(out.cpu().numpy(), should_output, 1e-2)
```

cpu version
截图 2023-01-14 10-18-47

pure python version
截图 2023-01-14 10-18-13


In fact, due to the volatility of benchmark results, there was not much difference between the two in my multiple tests, and the cpu version had a slight advantage over the python version

Comment (grimoire) [reply mmcv/ops/csrc/pytorch/cpu/bbox_overlaps_cpu.cpp]: cool

**Review by zhouzaida — APPROVED** [review]
